### PR TITLE
bump-formula-pr: make it usable on linux

### PR DESCRIPTION
**Pull request by dawidd6**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Previous effort (looks stale): closes https://github.com/Homebrew/brew/pull/6538

This commit makes it possible to use `bump-formula-pr` on linux to
contribute to `homebrew-core` as easily as it is on macOS, but with some
additional steps made before actual bumping of the formula.

Differences in the process (before bumping):
- add `homebrew-core` remote named `homebrew`
- fetch `homebrew/master`
- check if formula exists in `homebrew-core`
- switch to `homebrew/master` ref

Here are the PRs created using `brew bump-formula-pr` on linux:
https://github.com/Homebrew/homebrew-core/pull/46424
https://github.com/Homebrew/homebrew-core/pull/46589
https://github.com/Homebrew/homebrew-core/pull/46757

### CODE REVIEW

**Comment by issyl0** [comment]
Thanks so much for picking this up!

I know this is only a draft, but could you please add more context to the commit messages as to what you've done and _why_?

I'll test the actual code out a bit later. 👍

**Comment by dawidd6** [comment]
Yep, will add what I wrote in PR description to the commit as well and probably even more once I make it more reviewable :D.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 1]
```diff
@@ -83,6 +83,32 @@ def bump_formula_pr
       checked_for_duplicates = true
     end
 
+    linux_core = OS.linux? && formula.tap.core_tap?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Might be worth pulling all this into another function?

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 2]
```diff
@@ -83,6 +83,32 @@ def bump_formula_pr
...
+        ohai "dry linux"
+      else
+        formula.path.parent.cd do
+          ohai "Fetching homebrew-core"
+          unless Utils.popen_read("git remote -v").match?(/^homebrew.*Homebrew\/homebrew-core(.git)?/)
+            safe_system "git", "remote", "add", "homebrew", "https://github.com/Homebrew/homebrew-core"
+          end
+          safe_system "git", "fetch", "homebrew", "master"
+          if system "git", "cat-file", "-e", "homebrew/master:#{formula.path.to_s[/(Formula\/.*)/, 1]}"
+            ohai "#{formula.full_name} does exist in homebrew-core"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]:
```suggestion
            ohai "#{formula.full_name} exists in homebrew-core!"
```

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 3]
```diff
@@ -83,6 +83,32 @@ def bump_formula_pr
...
+            previous_branch = Utils.popen_read("git symbolic-ref -q --short HEAD").chomp
+            safe_system "git", "checkout", "homebrew/master"
+          else
+            ohai "#{formula.full_name} does not exist in homebrew-core (assuming it's linux-specific)"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]:
```suggestion
            ohai "#{formula.full_name} does not exist in homebrew-core so assuming formula is linuxbrew-core-specific"
```

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 1 above)
Comment (dawidd6) [reply Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Yes, I thought about that too. Will surely do!

Comment (dawidd6) [reply Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Done initially, but don't know if it's properly to the end.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 4]
```diff
@@ -66,6 +66,37 @@ def bump_formula_pr_args
     end
   end
 
+  def handle_linux(formula)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]:
```suggestion
  def use_correct_linux_tap(formula)
```

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 5]
```diff
@@ -83,6 +114,10 @@ def bump_formula_pr
       checked_for_duplicates = true
     end
 
+    if OS.linux? && formula.tap.core_tap?
+      tap, origin_branch, previous_branch = handle_linux(formula)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]:
```suggestion
      tap, origin_branch, previous_branch = use_correct_linux_tap(formula)
    else
      tap = formula.tap.full_name
      origin_branch = "origin/master"
      previous_branch = "-"
```

If these can't be `nil` from `use_correct_linux_tap`/`handle_linux`.

If they can maybe doing the `||=` directly below would be nicer.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 6]
```diff
@@ -83,6 +114,9 @@ def bump_formula_pr
       checked_for_duplicates = true
     end
 
+    # Must be done earlier
+    tap, origin_branch, previous_branch = handle_linux(formula) if OS.linux? && formula.tap.core_tap?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Could even check `if OS.linux? && formula.tap.core_tap?` in the method?

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 5 above)
Comment (dawidd6) [reply Library/Homebrew/dev-cmd/bump-formula-pr.rb]: I've put an early return in the function itself with default values.

**Comment by dawidd6** [comment]
Would be nice to have some bumpable formula that I can test this PR on.

**Comment by dawidd6** [comment]
Also one problem that I encountered and don't have an idea how to solve is a situation where `brew bump-formula-pr` exits if a PR for the formula is already open and we are left on `homebrew/master` branch in core tap. Would need to implement some kind of failsafe to revert back to previous branch.

**Comment by dawidd6** [comment]
Another successful PR: https://github.com/Homebrew/homebrew-core/pull/46757

**Comment by MikeMcQuaid** [comment]
Thanks again @dawidd6!

**Comment by issyl0** [comment]
Hurray! 🎉

**Pull request merged**